### PR TITLE
Add activities history logs

### PR DESCRIPTION
**Pull request by cp-amisha-i**
- [x] Add sub collection activity inside users collection 
- [x] Removed add expense tab and add activity tab & add + expense button in group list
- [x] Add is_active field for restore group / expense and transaction and manage restore 
- [x] Add updated_by field in group / expense and transaction & update it properly
- [x] Add all users activity logs when user added / updated / deleted group / expense and transaction
- [x] Implement activity screen and apply pagination, pull to refresh and add date header for every day logs 
- [x] Add Restore option in detail screen to restore group / expense / transaction and for group add restore view in group home
- [x] Deactivate all expenses and transactions when group is deleted to restore it in future.
- [x] Fix: expense list & scroll to top btn not move up with keyboard in group home screen 

https://github.com/user-attachments/assets/2aaa8e50-53ac-4a8b-a63a-ea07e92475f8    

 Simulator Screenshot - iPhone 16 - 2024-10-24 at 12 05 51 (1)

## Release Notes

- **New Features**
  - Introduced activity logging for expenses, transactions, and group activities.
  - Added new views for displaying activity logs and managing expenses.
  - Enhanced date handling with new methods for date comparisons and formatting.
  - New localization strings to improve user notifications and interactions.
  - Added a `RestoreButton` component for restoring deleted items.

- **Bug Fixes**
  - Improved error handling across various views and models.
  - Resolved issues with toolbar visibility based on state changes.

- **Enhancements**
  - Streamlined group management functionalities, including member actions and activity logging.
  - Improved UI for better user interaction and experience in managing expenses and activities.
  - Refined focus management for search bars and button visibility.

- **Assets**
  - Added new icons and images for improved visual representation in the app.

### CODE REVIEW

File: Splito/UI/Home/Expense/Expense Detail/ExpenseDetailsViewModel.swift [hunk 1]
```diff
@@ -117,12 +164,26 @@ class ExpenseDetailsViewModel: BaseViewModel, ObservableObject {
...
         Task {
             do {
                 viewState = .loading
-                try await expenseRepository.deleteExpense(groupId: groupId, expenseId: expenseId)
+                expense.updatedBy = userId
+
+                try await expenseRepository.deleteExpense(groupId: groupId, expense: expense)
                 NotificationCenter.default.post(name: .deleteExpense, object: expense)
                 await self.updateGroupMemberBalance(updateType: .Delete)
+                await addLogForDeletedExpense(type: .expenseDeleted)
```
Comment (cp-amisha-i) [inline Splito/UI/Home/Expense/Expense Detail/ExpenseDetailsViewModel.swift:185]: Is it possible to call all this log adding methods from the respective repository rather than ViewModel, and also this log adding operation should not be part of sequential calls, it should work parellaly without blocking any other operation, as adding or updating any detail has nothing to do with the logs.

File: Splito/UI/Home/Expense/Expense Detail/ExpenseDetailsView.swift [hunk 2]
```diff
@@ -169,3 +177,16 @@ private struct ExpenseInfoView: View {
         }
     }
 }
+
+struct RestoreButton: View {
```
Comment (cp-amisha-i) [inline Splito/UI/Home/Expense/Expense Detail/ExpenseDetailsView.swift:181]: Move this function to base as its reusing.

File: Splito/UI/Home/Expense/Expense Detail/ExpenseDetailsViewModel.swift [hunk 3]
```diff
@@ -103,9 +105,54 @@ class ExpenseDetailsViewModel: BaseViewModel, ObservableObject {
...
+                         message: "This expense could not be edited. You do not have permission to edit this expense, Sorry!")
+            return
+        }
+
         showEditExpenseSheet = true
     }
 
+    func handleRestoreButtonAction() {
+        showAlert = true
+        alert = .init(title: "Undelete expense",
```
Comment (cp-amisha-i) [inline Splito/UI/Home/Expense/Expense Detail/ExpenseDetailsViewModel.swift:119]: Typo `Undelete` -> `Restore`

File: Splito/UI/Home/Expense/Expense Detail/ExpenseDetailsViewModel.swift [hunk 4]
```diff
@@ -117,12 +164,26 @@ class ExpenseDetailsViewModel: BaseViewModel, ObservableObject {
...
+
         Task {
             do {
                 viewState = .loading
-                try await expenseRepository.deleteExpense(groupId: groupId, expenseId: expenseId)
+                expense.updatedBy = userId
+
+                try await expenseRepository.deleteExpense(groupId: groupId, expense: expense)
                 NotificationCenter.default.post(name: .deleteExpense, object: expense)
                 await self.updateGroupMemberBalance(updateType: .Delete)
```
Comment (cp-amisha-i) [inline Splito/UI/Home/Expense/Expense Detail/ExpenseDetailsViewModel.swift:184]: Move these 2 functions to repository and try to call them without blocking the main operation.

File: Splito/UI/Home/Groups/Group/GroupHomeViewModel.swift [hunk 5]
```diff
@@ -37,10 +38,10 @@ class GroupHomeViewModel: BaseViewModel, ObservableObject {
     @Published var showSimplifyInfoSheet = false
     @Published var showInviteMemberSheet = false
 
-    @Published private(set) var showSearchBar = false
-    @Published private(set) var showScrollToTopBtn = false
+    @Published var showSearchBar = false
+    @Published var showScrollToTopBtn = false
```
Comment (cp-amisha-i) [inline Splito/UI/Home/Groups/Group/GroupHomeViewModel.swift:42]: Why did you remove private(set) from all variables?

File: expense-notifications-functions/lib/index.js [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+Object.defineProperty(exports, "__esModule", { value: true });
+const firestore_1 = require("firebase-admin/firestore");
+const firestore_2 = require("firebase-functions/v2/firestore");
+const admin = require("firebase-admin");
+const logger = require("firebase-functions/logger");
+admin.initializeApp();
+const db = (0, firestore_1.getFirestore)();
+const notificationTitle = `Splito`;
+// Cloud Function to handle new expense creation and notify users
+exports.onExpenseCreated = (0, firestore_2.onDocumentCreated)({ document: 'groups/{groupId}/expenses/{expenseId}' }, async (event) => {
```
Comment (cp-amisha-i) [inline expense-notifications-functions/lib/index.js:12]: Why did we add this?

File: Data/Data/Model/Expense.swift [hunk 7]
```diff
@@ -16,20 +16,24 @@ public struct Expense: Codable, Hashable {
     public var date: Timestamp
     public var paidBy: [String: Double]
     public let addedBy: String
+    public var updatedBy: String
```
[the remark this answers is not in the document]
Comment (cp-nirali-s) [reply Data/Data/Model/Expense.swift:19]: I will add this in push notification branch as this mr contains so much changes so, i already added this in my queue

File: Data/Data/Model/ActivityLog.swift [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    @DocumentID public var id: String? // Automatically generated ID by Firestore
+
+    public let type: ActivityType   // The type of activity (e.g., group created, expense updated)
+    public let groupId: String
+    public let activityId: String   // The id of the activity (e.g., expense or transaction)
+    public let groupName: String
+    public let actionUserName: String   // The name of the user who performed the action
+    public let recordedOn: Timestamp
+    public let previousGroupName: String?    // The previous name of the group
+    public let removedMemberName: String?
```
[the remark this answers is not in the document]
Comment (cp-nirali-s) [reply Data/Data/Model/ActivityLog.swift:26]: You removed yourself from the group "Test Log".

File: Splito/UI/Home/Account/AccountHomeView.swift [hunk 9]
```diff
@@ -56,9 +55,6 @@ struct AccountHomeView: View {
             }
         }
         .background(surfaceColor)
-        .onAppear {
-            homeRouteViewModel.updateSelectedGroup(id: nil)
```
[the remark this answers is not in the document]
Comment (cp-nirali-s) [reply Splito/UI/Home/Account/AccountHomeView.swift:60]: Previously we are calling ExpenseRouteView() in tabview but now we have added button in group home and we don't have any expense tab so we can directly pass selected group id to ExpenseRouteView() in group home.

File: Splito/UI/Home/ActivityLog/ActivityLogViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        formatter.dateFormat = "dd MMMM yyyy"
+        return formatter
+    }()
+
+    init(router: Router<AppRoute>) {
+        self.router = router
+        super.init()
+
+        self.fetchActivityLogsInitialData()
+        self.fetchLatestActivityLogs()
```
[the remark this answers is not in the document]
Comment (cp-nirali-s) [reply Splito/UI/Home/ActivityLog/ActivityLogViewModel.swift:41]: We decided but if we will do manual refresh then we need to manage 16 cases from different viewmodels and in activity view model we need to observe those 16 cases, so code size will increase too much, i checked and it's works fine with real time updates because i only add those log in activity log array which is new.

File: Splito/UI/Home/Expense/AddExpenseViewModel.swift [hunk 11]
```diff
@@ -340,12 +343,35 @@ extension AddExpenseViewModel {
...
+                groupTasks.addTask { [weak self] in
+                    await self?.addActivityLog(expense: expense, type: .expenseAdded, memberId: memberId, errors: &errors)
+                }
+            }
+        }
+
+        handleActivityLogErrors(errors)
+    }
+
     private func updateExpense(groupId: String, expense: Expense, oldExpense: Expense, completion: (Bool) -> Void) async {
```
[the remark this answers is not in the document]
Comment (cp-nirali-s) [reply Splito/UI/Home/Expense/AddExpenseViewModel.swift:366]: Yes, they are allowing and i have added same behaviour wherever possible.

File: Splito/UI/Home/Groups/Group/GroupHomeViewModel.swift (hunk 5 above)
Comment (cp-nirali-s) [reply Splito/UI/Home/Groups/Group/GroupHomeViewModel.swift:42]: Because i am getting file length violation error, so i need to create extension of GroupHomeViewModel in different file and in different file we can't modify those variable which marked as private(set), we can modify those variables in same file only so...
